File: Programming Language Concepts/HW/7/collcount.fs
\ collcount.fs
\ Jon Rippe
\ CSCE 331 - Spring 2021
\ HW 7


\ n is pos integer & c is number of iterations of the
\ Collatz function required to take n to 1
: collcount { n -- c }

    0 { c }

    begin
    n 1 <> while
        n 2 mod 0 = if
            n 2 / to n
        else
            n 3 * 1 + to n
        endif
        c 1 + to c
    repeat

    c
;

File: Programming Language Concepts/HW/7/collcount.pl
% collcount.pl
% Jon Rippe
% CSCE 331 - Spring 2021
% HW 7


% n is pos integer & c is number of iterations of the
% Collatz function required to take n to 1
collcount(1, C) :- C is 0, !.

collcount(N, C) :-
    0 is mod(N, 2),
    N0 is N/2,
    collcount(N0, C0),
    C is C0+1, !.

collcount(N, C) :-
    N0 is (N*3)+1,
    collcount(N0, C0),
    C is C0+1.